File: Python/sorting/bubble_sort.py
"""
Bubble Sort Algorithm

Time Complexity: O(n^2) in worst and average case, O(n) in best case (when array is already sorted)
Space Complexity: O(1)

Bubble sort is a simple sorting algorithm that repeatedly steps through the list,
compares adjacent elements and swaps them if they are in the wrong order.
"""

def bubble_sort(arr):
    """
    Sorts an array using bubble sort algorithm.
    
    Args:
        arr: List of comparable elements to be sorted
    
    Returns:
        Sorted list (modifies the input list in-place and returns it)
    
    Example:
        >>> bubble_sort([64, 34, 25, 12, 22, 11, 90])
        [11, 12, 22, 25, 34, 64, 90]
    """
    n = len(arr)
    
    # Traverse through all array elements
    for i in range(n):
        # Last i elements are already in place
        swapped = False
        
        for j in range(0, n - i - 1):
            # Swap if the element found is greater than the next element
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        
        # If no swapping happened, array is already sorted
        if not swapped:
            break
    
    return arr


def bubble_sort_descending(arr):
    """
    Sorts an array in descending order using bubble sort.
    
    Args:
        arr: List of comparable elements to be sorted
    
    Returns:
        Sorted list in descending order
    
    Example:
        >>> bubble_sort_descending([64, 34, 25, 12, 22, 11, 90])
        [90, 64, 34, 25, 22, 12, 11]
    """
    n = len(arr)
    
    for i in range(n):
        swapped = False
        
        for j in range(0, n - i - 1):
            # Swap if the element found is smaller than the next element
            if arr[j] < arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        
        if not swapped:
            break
    
    return arr


if __name__ == "__main__":
    # Example usage
    print("Bubble Sort Examples:")
    
    # Test ascending sort
    arr1 = [64, 34, 25, 12, 22, 11, 90]
    print(f"\nOriginal array: {arr1}")
    sorted_arr1 = bubble_sort(arr1.copy())
    print(f"Sorted array (ascending): {sorted_arr1}")
    
    # Test descending sort
    arr2 = [64, 34, 25, 12, 22, 11, 90]
    sorted_arr2 = bubble_sort_descending(arr2.copy())
    print(f"Sorted array (descending): {sorted_arr2}")
    
    # Test with already sorted array
    arr3 = [1, 2, 3, 4, 5]
    print(f"\nAlready sorted array: {arr3}")
    sorted_arr3 = bubble_sort(arr3.copy())
    print(f"After bubble sort: {sorted_arr3}")
    
    # Test with single element
    arr4 = [42]
    print(f"\nSingle element array: {arr4}")
    sorted_arr4 = bubble_sort(arr4.copy())
    print(f"After bubble sort: {sorted_arr4}")

File: Java/sorting/BubbleSort.java
/**
 * Bubble Sort Algorithm
 * 
 * Time Complexity: O(n^2) in worst and average case, O(n) in best case (when array is already sorted)
 * Space Complexity: O(1)
 * 
 * Bubble sort is a simple sorting algorithm that repeatedly steps through the list,
 * compares adjacent elements and swaps them if they are in the wrong order.
 */
public class BubbleSort {
    
    /**
     * Sorts an array using bubble sort algorithm (ascending order).
     * 
     * @param arr Array of integers to be sorted
     * @return Sorted array
     */
    public static int[] bubbleSort(int[] arr) {
        int n = arr.length;
        
        // Traverse through all array elements
        for (int i = 0; i < n; i++) {
            boolean swapped = false;
            
            // Last i elements are already in place
            for (int j = 0; j < n - i - 1; j++) {
                // Swap if the element found is greater than the next element
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }
            
            // If no swapping happened, array is already sorted
            if (!swapped) {
                break;
            }
        }
        
        return arr;
    }
    
    /**
     * Sorts an array in descending order using bubble sort.
     * 
     * @param arr Array of integers to be sorted
     * @return Sorted array in descending order
     */
    public static int[] bubbleSortDescending(int[] arr) {
        int n = arr.length;
        
        for (int i = 0; i < n; i++) {
            boolean swapped = false;
            
            for (int j = 0; j < n - i - 1; j++) {
                // Swap if the element found is smaller than the next element
                if (arr[j] < arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }
            
            if (!swapped) {
                break;
            }
        }
        
        return arr;
    }
    
    /**
     * Utility method to print an array
     */
    public static void printArray(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
    
    /**
     * Main method with example usage
     */
    public static void main(String[] args) {
        System.out.println("Bubble Sort Examples:");
        
        // Test ascending sort
        int[] arr1 = {64, 34, 25, 12, 22, 11, 90};
        System.out.print("\nOriginal array: ");
        printArray(arr1);
        
        bubbleSort(arr1);
        System.out.print("Sorted array (ascending): ");
        printArray(arr1);
        
        // Test descending sort
        int[] arr2 = {64, 34, 25, 12, 22, 11, 90};
        bubbleSortDescending(arr2);
        System.out.print("Sorted array (descending): ");
        printArray(arr2);
        
        // Test with already sorted array
        int[] arr3 = {1, 2, 3, 4, 5};
        System.out.print("\nAlready sorted array: ");
        printArray(arr3);
        bubbleSort(arr3);
        System.out.print("After bubble sort: ");
        printArray(arr3);
        
        // Test with single element
        int[] arr4 = {42};
        System.out.print("\nSingle element array: ");
        printArray(arr4);
        bubbleSort(arr4);
        System.out.print("After bubble sort: ");
        printArray(arr4);
    }
}
